server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java

```diff
@@ -365,31 +363,31 @@ boolean moveToTrash(Path path) throws IOException {
   private void getZooLock(HostAndPort addr) throws KeeperException, InterruptedException {
     var path = ServiceLock.path(getContext().getZooKeeperRoot() + Constants.ZGC_LOCK);
 
-    LockWatcher lockWatcher = new LockWatcher() {
-      @Override
-      public void lostLock(LockLossReason reason) {
-        Halt.halt("GC lock in zookeeper lost (reason = " + reason + "), exiting!", 1);
-      }
+    UUID zooLockUUID = UUID.randomUUID();
+    gcLock = new ServiceLock(getContext().getZooReaderWriter().getZooKeeper(), path, zooLockUUID);
 
-      @Override
-      public void unableToMonitorLockNode(final Exception e) {
-        // ACCUMULO-3651 Level changed to error and FATAL added to message for slf4j compatibility
-        Halt.halt(-1, () -> log.error("FATAL: No longer able to monitor lock node ", e));
+    while (true) {
+      HAServiceLockWatcher gcLockWatcher = new HAServiceLockWatcher("gc");
+      gcLock.lock(gcLockWatcher,
+          new ServerServices(addr.toString(), Service.GC_CLIENT).toString().getBytes(UTF_8));
 
+      gcLockWatcher.waitForChange();
+
+      if (gcLockWatcher.isLockAcquired()) {
+        break;
       }
-    };
 
-    UUID zooLockUUID = UUID.randomUUID();
-    gcLock = new ServiceLock(getContext().getZooReaderWriter().getZooKeeper(), path, zooLockUUID);
-    while (true) {
-      if (gcLock.tryLock(lockWatcher,
-          new ServerServices(addr.toString(), Service.GC_CLIENT).toString().getBytes(UTF_8))) {
-        log.debug("Got GC ZooKeeper lock");
-        return;
+      if (!gcLockWatcher.isFailedToAcquireLock()) {
+        throw new IllegalStateException("gc lock in unknown state");
       }
-      log.debug("Failed to get GC ZooKeeper lock, will retry");
-      sleepUninterruptibly(1, TimeUnit.SECONDS);
+
+      gcLock.tryToCancelAsyncLockOrUnlock();
+
+      sleepUninterruptibly(1000, TimeUnit.MILLISECONDS);
```

Comment:
Could keep this debug.

```suggestion
      log.debug("Failed to get GC ZooKeeper lock, will retry");
      sleepUninterruptibly(1000, TimeUnit.MILLISECONDS);
```

Reply:
Good catch, I didn't realize I got rid of that.